exonum/src/node/mod.rs

```diff
@@ -794,6 +797,35 @@ pub struct ConnectInfo {
     pub public_key: PublicKey,
 }
 
+impl ConnectInfo {
+    /// Tries to serialize the given `ConnectInfo` into a UTF-8 encoded JSON.
+    pub fn try_serialize(self) -> serde_json::Result<Vec<u8>> {
+        serde_json::to_vec(&self)
+    }
+
+    /// Tries to deserialize the given `value` into `ConnectInfo`.
+    pub fn try_deserialize(value: &[u8]) -> serde_json::Result<Self> {
+        serde_json::from_slice(value)
+    }
+}
+
+impl StorageValue for ConnectInfo {
```

Comment:
You should implement `ConnectInfo` via `encoding_struct` and also implement some traits. See code for example.
https://github.com/alekseysidorov/exonum-btc-anchoring/blob/native-segwit/src/btc/macros.rs#L98

Reply:
I tried, but `encoding_struct` doesn't work with `deserialize_with`.

Reply:
Why you need to use `deserialize_with`?

Reply:
We use `deserialize_with=deserialize_socket_address` to check validity of `SocketAddr`.

Reply:
Off-topic: `deserialize_with` is used not to check validity, but rather to resolve hosts during deserialization. Which, IMO, looks like magic overdose (host resolution is non-deterministic and involves external global state); I'd be more satisfied with later resolution separated from deserialization. Btw, the present deserialization most probably does not survive a round-trip if used with the [non-human-readable](https://docs.rs/serde/1.0.72/serde/trait.Serializer.html#method.is_human_readable) / binary format.

Reply:
To add to the previous comment: If I understand correctly, during serialization we _always_ record a resolved `SocketAddr` (I'm not saying it's necessarily correct in all cases), so during deserialization from storage we don't need custom resolution logic (= `deserialize_with`). The whole process looks significantly 🤦‍♂️ to me, so I'd suggest to at least explore the possibility to use choose the type for the peer address more smartly. As a possibility: use `{ host: String, port: u16 }` tuple (cf. [`ToSocketAddrs` implementation for `(&str, u16)`](https://doc.rust-lang.org/src/std/net/addr.rs.html#865-882)) and resolve it *after* deserialization.

Reply:
You are right when deserializing from storage when don't need `deserialize_with`. But we also use `ConnectInfo` in config files and `deserialize_with` added only to resolve addresses when parsing config. It may be worth to use different `ConnectInfo` for storage and config files.

Reply:
I'm not sure using a resolved address in stored `ConnectInfo` is the correct way to proceed in the long run:

- A host may be resolved to multiple addresses, some of which may not work for any particular node unpredictably.
- The resolution may change with time for a multitude of reasons.

Reply:
@slowli Such approach for resolving addresses was implemented as a first step. We've planed to change it in the near future. More correct solution is to resolve addresses before connect to the nodes. But in this case we should change network messages cause they contain `SocketAddr` rather then `String` for address of peer record.

Reply:
Do we have a task to track this issue with resolving?

Reply:
@ilammy yes, https://jira.bf.local/browse/ECR-2310